source/common/network/io_uring_socket_handle_impl.cc

```diff
@@ -0,0 +1,415 @@
+  ENVOY_LOG(trace, "writev, fd = {}, type = {}", fd_, ioUringSocketTypeStr());
+
+  auto write_result = checkWriteResult();
+  if (write_result.has_value()) {
+    return std::move(*write_result);
+  }
+
+  uint64_t ret = io_uring_socket_->write(slices, num_slice);
+  return {ret, IoSocketError::none()};
+}
+
+Api::IoCallUint64Result IoUringSocketHandleImpl::write(Buffer::Instance& buffer) {
+  ENVOY_LOG(trace, "write {}, fd = {}, type = {}", buffer.length(), fd_, ioUringSocketTypeStr());
+
+  auto write_result = checkWriteResult();
+  if (write_result.has_value()) {
+    return std::move(*write_result);
+  }
+
+  uint64_t buffer_size = buffer.length();
+  io_uring_socket_->write(buffer);
+  return {buffer_size, IoSocketError::none()};
+}
+
+Api::IoCallUint64Result IoUringSocketHandleImpl::sendmsg(const Buffer::RawSlice*, uint64_t, int,
+                                                         const Address::Ip*,
+                                                         const Address::Instance&) {
+  ENVOY_LOG(trace, "sendmsg, fd = {}, type = {}", fd_, ioUringSocketTypeStr());
+  return Network::IoSocketError::ioResultSocketInvalidAddress();
+}
+
+Api::IoCallUint64Result IoUringSocketHandleImpl::recvmsg(Buffer::RawSlice*, const uint64_t,
+                                                         uint32_t, RecvMsgOutput&) {
+  ENVOY_LOG(trace, "recvmsg, fd = {}, type = {}", fd_, ioUringSocketTypeStr());
+  return Network::IoSocketError::ioResultSocketInvalidAddress();
+}
+
+Api::IoCallUint64Result IoUringSocketHandleImpl::recvmmsg(RawSliceArrays&, uint32_t,
+                                                          RecvMsgOutput&) {
+  ENVOY_LOG(trace, "recvmmsg, fd = {}, type = {}", fd_, ioUringSocketTypeStr());
+  return Network::IoSocketError::ioResultSocketInvalidAddress();
+}
+
+Api::IoCallUint64Result IoUringSocketHandleImpl::recv(void* buffer, size_t length, int flags) {
+  ASSERT(io_uring_socket_.has_value());
+  ENVOY_LOG(trace, "recv, fd = {}, type = {}", fd_, ioUringSocketTypeStr());
+
+  // The only used flag in Envoy is MSG_PEEK for listener filters, including TLS inspectors.
+  ASSERT(flags == 0 || flags == MSG_PEEK);
+  Buffer::RawSlice slice;
+  slice.mem_ = buffer;
+  slice.len_ = length;
+  if (flags == 0) {
+    return readv(length, &slice, 1);
+  }
+
+  return copyOut(length, &slice, 1);
+}
+
+Api::SysCallIntResult IoUringSocketHandleImpl::bind(Address::InstanceConstSharedPtr address) {
+  ENVOY_LOG(trace, "bind {}, fd = {}, io_uring_socket_type = {}", address->asString(), fd_,
+            ioUringSocketTypeStr());
+  return Api::OsSysCallsSingleton::get().bind(fd_, address->sockAddr(), address->sockAddrLen());
+}
+
+Api::SysCallIntResult IoUringSocketHandleImpl::listen(int backlog) {
+  ENVOY_LOG(trace, "listen, fd = {}, type = {}", fd_, ioUringSocketTypeStr());
+
+  ASSERT(io_uring_socket_type_ == IoUringSocketType::Unknown);
+
+  io_uring_socket_type_ = IoUringSocketType::Accept;
```

Comment:
We won't implement accept socket with iouring since there is race issue between worker thread and main thread need to be resolved.

so we should use the shadow io handle here to ensure the accept socket still using the socket API.
https://github.com/envoyproxy/envoy/pull/24958/files#diff-6c6828df3b66432a238d5395edcc5ee9845d08d218977967dfc9abc9faf87050R309

Reply:
Sockets for accept have been fallen back to the default socket API behavior. I just do not like adding `shadow_io_handle` for debugging in a formal patch, but we can definitely put their in your `io_uring_test` branch.

Reply:
ok, I saw you fallback to default socket at accept method.